onair/data_handling/sbn_adapter.py

```diff
@@ -0,0 +1,198 @@
+
+        read_index = 0
+        with self.new_data_lock:
+            self.new_data = False
+            self.double_buffer_read_index = (self.double_buffer_read_index + 1) % 2
+            read_index = self.double_buffer_read_index
+
+        return self.currentData[read_index]['data']
+
+    def has_more(self):
+        """Returns true if the adapter has more data.
+           For now always true: connection should be live as long as cFS is running.
+           TODO: allow to detect if cFS/the connection has died"""
+        return True
+
+    def message_listener_thread(self):
+        """Thread to listen for incoming messages from SBN"""
+
+        while(True):
+            generic_recv_msg_p = POINTER(sbn.sbn_data_generic_t)()
+            sbn.recv_msg(generic_recv_msg_p)
+
+            msgID = generic_recv_msg_p.contents.TlmHeader.Primary.StreamId
+            app_name, data_struct = self.msgID_lookup_table[msgID]
+
+            recv_msg_p = POINTER(data_struct)()
+            recv_msg_p.contents = generic_recv_msg_p.contents
+            recv_msg = recv_msg_p.contents
+
+            # prints out the data from the message to the terminal
+            print(", ".join([field_name + ": " + str(getattr(recv_msg, field_name)) for field_name, field_type in recv_msg._fields_[1:]]))
+
+            # TODO: Lock needed here?
+            self.get_current_data(recv_msg, data_struct, app_name)
+
+    def get_current_data(self, recv_msg, data_struct, app_name):
+        # TODO: Lock needed here?
+        current_buffer = self.currentData[(self.double_buffer_read_index + 1) %2]
+        secondary_header = recv_msg.TlmHeader.Secondary
+        #gets seconds from header and adds to current buffer
+        start_time = datetime.datetime(1969, 12, 31, 20) # January 1, 1980
+        seconds = secondary_header.Seconds
+        subseconds = secondary_header.Subseconds
+        curr_time = seconds + (2**(-32) * subseconds) # a subsecond is equal to 2^-32 second
+        time = start_time + datetime.timedelta(seconds=curr_time)
+        str_time = time.strftime("%Y-%j-%H:%M:%S.%f")
+        current_buffer['data'][0] = str_time
+
+        # Skip the header, walk through the stuct
+        for field_name, field_type in recv_msg._fields_[1:]:
+            field_names = self.gather_field_names(field_name, field_type)
+
+            for name in field_names:
+                idx = current_buffer['headers'].index(app_name + "." + name)
+                # Pull the data out of the message buy walking down the nested types
+                data = ""
+                current_object = recv_msg
+                for sub_type in name.split('.'):
+                    current_object = getattr(current_object, sub_type)
+                    data = str(current_object) # note does not work for arrays?
+                current_buffer['data'][idx] = data
+
+        with self.new_data_lock:
+            self.new_data = True
+
+    def has_data(self):
+        return self.new_data
```

Comment:
Please leave newlines at the end of files (it makes it look nicer when someone cats a file in their terminal)